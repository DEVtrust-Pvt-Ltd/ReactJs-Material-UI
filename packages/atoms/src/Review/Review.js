import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { Box, makeStyles } from "@material-ui/core";
import Rating from "@material-ui/lab/Rating";
import { colors } from "@cxa-shop-ui/styles";

const useStyles = makeStyles((theme) => ({
    container: {
        position: "relative",
        borderBottom: `1px solid ${colors.veryLightGray}`,
        paddingTop: theme.spacing(1.25),
        paddingBottom: theme.spacing(1.25),
        paddingLeft: theme.spacing(0),
        paddingRight: theme.spacing(0),
    },
    author: {
        textTransform: "capitalize",
        fontWeight: 600,
        fontSize: "16px",
        color: colors.mediumEmphasis,
        marginBottom: theme.spacing(0),
    },
    date: {
        fontWeight: 600,
        fontSize: "14px",
        color: colors.mediumEmphasis,
    },
    message: {
        fontWeight: 450,
        width: '100px',
        fontSize: "14px",
        color: colors.mediumEmphasis,
        marginBottom: theme.spacing(1.25),
        marginTop: theme.spacing(0),
    },
    read_more: {
        fontWeight: 600,
        fontSize: "16px",
        color: colors.mediumEmphasis,
    },
    rateWrap: {
        display: 'flex',
        justifyContent: 'space-between',
        flexDirection: 'row',
        paddingBottom: theme.spacing(0.7),
    },
    rating: {
        fontSize: '16px',
        color: colors.grey2,
        marginTop: theme.spacing(1.25),
        marginBottom: theme.spacing(1.25),
        marginRight: theme.spacing(0),
        marginLeft: theme.spacing(0),
    },
}));

const Review = ({
    author,
    date,
    message,
    charLimit,
    readMoreText,
    hideFullText,
    maxRating,
    rating,
    precision,
    disabled,
}) => {
    const [open, setOpen] = useState(false);
    const [value, setValue] = React.useState(rating);

    useEffect(() => setValue(rating));

    const classes = useStyles();

    const finalMessage = () => {
        return message.length > charLimit && !open
            ? message.slice(0, charLimit) + "..."
            : message;
    };

    const showButton = () => {
        return message.length > charLimit;
    };

    const toggleMessage = () => {
        setOpen((state) => !state);
    };

    const buttonText = () => {
        let text = readMoreText;

        if (open) {
            text = hideFullText;
        }
        return text;
    };

    return (
        <Box className={classes.container}>
            <Box className={classes.rateWrap}>
                {author && <span className={classes.author}>{author}</span>}
                <span className={classes.date}>{date}</span>
            </Box>

            {rating > 0 && (
                <Box>
                    <Rating
                        name="review"
                        max={maxRating}
                        precision={precision}
                        value={value}
                        onChange={(event, newValue) => {
                            setValue(newValue);
                        }}
                        disabled={disabled}
                        className={classes.rating}
                    />
                </Box>
            )}

            {message && (
                <Box>
                    <span className={classes.message}>
                        {finalMessage()} &nbsp;
                        {showButton() && (
                            <span onClick={toggleMessage} className={classes.read_more}>
                                {buttonText()}
                            </span>
                        )}
                    </span>
                </Box>
            )}
        </Box>
    );
};

Review.propTypes = {
    message: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    author: PropTypes.string,
    date: PropTypes.string,
    disabled: PropTypes.bool,
    rating: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.bool,
    ]),
    maxRating: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    precision: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    charLimit: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.bool,
    ]),
    readMoreText: PropTypes.string,
    hideFullText: PropTypes.string,
};

Review.defaultProps = {
    /**
     * Author of the review
     */
    author: "",
    /**
     * Date of the review
     */
    date: "",
    /**
     * Message from the reviewer
     */
    message: "",
    /**
     * Rating from the reviewer
     */
    rating: false,

    /**
     * Rating from the reviewer
     */
    disabled: true,
    /**
     * Precision of the review
     */
    precision: 0.5,
    /**
     * Max rating for the review
     */
    maxRating: 5,
    /**
     * Char limit for the review
     */
    charLimit: 250,
    /**
     * Read more text for the review
     */
    readMoreText: "Read more",
    /**
     * Hide full text message for the review
     */
    hideFullText: "Read less",
};

export default Review;
