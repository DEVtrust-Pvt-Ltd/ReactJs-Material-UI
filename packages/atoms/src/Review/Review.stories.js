import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, number, text, boolean } from "@storybook/addon-knobs";

import Review from ".";

storiesOf("Atoms|Review", module)
    .addDecorator(withKnobs)
    .add("Primary", () => (
        <Review
            author={text("author (prop)", "Author name")}
            date={text("date (prop)", "Sep 2019")}
            message={text(
                "message (prop)",
                "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged."
            )}
            disabled={boolean("disabled (prop)", true)}
            rating={number("rating (prop)", 4)}
            maxRating={number("maxRating (prop)", 5)}
            precision={number("precision (prop)", 0.5)}
            charLimit={number("charLimit (prop)", 250)}
            readMoreText={text("readMoreText (prop)", "Read more")}
            hideFullText={text("hideFullText (prop)", "Read less")}
        />
    ));
