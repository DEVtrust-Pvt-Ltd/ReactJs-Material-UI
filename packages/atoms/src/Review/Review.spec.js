import React from "react";
import {render} from "@testing-library/react";
import Review from ".";

describe("Review Component", () => {
    it("should match snapshot when properties are present", () => {
        const {container} = render(
            <Review
                author="Author name"
                date="Sep 2019"
                message="Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged."
                rating={4}
                charLimit={250}
            />
        );

        expect(container).toMatchSnapshot();
    });
});
