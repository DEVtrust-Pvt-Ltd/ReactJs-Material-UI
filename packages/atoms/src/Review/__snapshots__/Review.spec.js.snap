// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review Component should match snapshot when properties are present 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-7 makeStyles-container-1"
  >
    <div
      class="MuiBox-root MuiBox-root-8 makeStyles-author-2"
    >
      Author name
    </div>
    <div
      class="MuiBox-root MuiBox-root-9"
    >
      <span
        class="MuiRating-root Mui-disabled"
      >
        <span
          class="MuiRating-decimal"
        >
          <label
            class="MuiRating-label"
            for="review-0-5"
            style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
          >
            <span
              class="MuiRating-icon MuiRating-iconFilled"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              0.5 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-0-5"
            name="review"
            type="radio"
            value="0.5"
          />
          <label
            class="MuiRating-label"
            for="review-1"
          >
            <span
              class="MuiRating-icon MuiRating-iconFilled"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              1 Star
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-1"
            name="review"
            type="radio"
            value="1"
          />
        </span>
        <span
          class="MuiRating-decimal"
        >
          <label
            class="MuiRating-label"
            for="review-1-5"
            style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
          >
            <span
              class="MuiRating-icon MuiRating-iconFilled"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              1.5 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-1-5"
            name="review"
            type="radio"
            value="1.5"
          />
          <label
            class="MuiRating-label"
            for="review-2"
          >
            <span
              class="MuiRating-icon MuiRating-iconFilled"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              2 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-2"
            name="review"
            type="radio"
            value="2"
          />
        </span>
        <span
          class="MuiRating-decimal"
        >
          <label
            class="MuiRating-label"
            for="review-2-5"
            style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
          >
            <span
              class="MuiRating-icon MuiRating-iconFilled"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              2.5 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-2-5"
            name="review"
            type="radio"
            value="2.5"
          />
          <label
            class="MuiRating-label"
            for="review-3"
          >
            <span
              class="MuiRating-icon MuiRating-iconFilled"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              3 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-3"
            name="review"
            type="radio"
            value="3"
          />
        </span>
        <span
          class="MuiRating-decimal"
        >
          <label
            class="MuiRating-label"
            for="review-3-5"
            style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
          >
            <span
              class="MuiRating-icon MuiRating-iconFilled"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              3.5 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-3-5"
            name="review"
            type="radio"
            value="3.5"
          />
          <label
            class="MuiRating-label"
            for="review-4"
          >
            <span
              class="MuiRating-icon MuiRating-iconFilled"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              4 Stars
            </span>
          </label>
          <input
            checked=""
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-4"
            name="review"
            type="radio"
            value="4"
          />
        </span>
        <span
          class="MuiRating-decimal"
        >
          <label
            class="MuiRating-label"
            for="review-4-5"
            style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
          >
            <span
              class="MuiRating-icon MuiRating-iconEmpty"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              4.5 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-4-5"
            name="review"
            type="radio"
            value="4.5"
          />
          <label
            class="MuiRating-label"
            for="review-5"
          >
            <span
              class="MuiRating-icon MuiRating-iconEmpty"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyhidden"
            >
              5 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyhidden"
            disabled=""
            id="review-5"
            name="review"
            type="radio"
            value="5"
          />
        </span>
      </span>
      <span
        class="makeStyles-date-3"
      >
        Sep 2019
      </span>
    </div>
    <div
      class="MuiBox-root MuiBox-root-35"
    >
      <p
        class="MuiTypography-root makeStyles-message-4 MuiTypography-body1"
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has...
          
        <span
          class="makeStyles-read_more-5"
        >
          Read more
        </span>
      </p>
    </div>
  </div>
</div>
`;
